# Práctica 6, Grupo 02 (Mario Calvarro, Beñat Pérez)

## Representación cromosomas
Para representar cada cromosoma hemos optado por una lista de 900 listas de seis enteros. Cada uno de estos enteros representa la cantidad de minas de un determinado tipo (van en orden) que se compran en ese segundo. 

## Población inicial
Para generar la población inicial hemos utilizado la siguiente asignación:

In [ ]:
number: int = int(random.randint(0, 1) * 1.045 **(i-90))
number = round(number * 3**(-6.5*ii))

Es decir, tiene un crecimiento exponencial respecto de los segundos e inversamente exponencial respecto de las minas. Por tanto, los últimos segundos se compraran muchas más minas y las minas de precio más bajo se compraran en mayor cantidad.

Los últimos 20 segundos no se compra ninguna mina.

## Operadores genéticos
### Selección
Para seleccionar que individuos se cruzaran o pasarán a la siguiente generación utilizamos una selección por torneo de grupos de 3 individuos donde tomamos aquel que tenga mayor fitness (oro al final de la partida).

### Mutación
Al final de la creación de una nueva generación, mutamos los individuos. La mutación de cada uno consiste en elegir aleatoriamente, con una proporción del 0.02%, segundos y sobre cada uno de estos elegidos multiplicar, con probabilidad, por 0.5 la cantidad de minas cambiadas. Como antes, la probabilidad crece exponencialmente respecto del segundo y decrece sobre el tipo de mina, es decir, los últimos segundos segundos cambiarán con mayor probabilidad la cantidad de minas compradas, siendo más probable este cambio para las minas baratas.  

### Cruze
Cruzamos cada individuo con su siguiente en la población. El cruze devuelve dos nuevos individuos y se realiza, sobre cada segundo, tomando el primer tercio y el último de un padre y el restante del otro. El padre con mayor cantidad de genes se elige aleatoriamente.

En base a un parámetro, una parte de la población pasa a la siguiente generación sin cruzarse.

## Función de evaluación
La función de evaluación consiste simplemente en la ejecucción del juego comprando minas según indica el cromosoma en sus genes. Cada vez que se realice una compra inválida, se múltiplica el fitness por 0.9 y se cambia el valor para que sea válido de nuevo.

## Parámetros
- Número de torneo: 3 individuos en los grupos del torneo.
- Número de generaciones: 200 (más generaciones no incrementa mucho el resultado final)
- Proporción de cruces: 30% (un valor mayor solo reduce el rendimiento)
- Tamaño de la población: 20 (de nuevo, un valor mayor solo reduce el rendimiento)

## Resultados
Como resultados finales hemos utilizado el siguiente script para realizar 100 ejecucciones y tomamos la media de ellas. Mostramos el número de dígitos de la solución porque al ser números tan grandes lo consideramos más representativo.

In [5]:
%%time
import math
import main

num_ej = 100
sum = 0
for i in range(num_ej):
    sum += main.res()

res = sum / num_ej
print(f"El número de dígitos de la solución media es {math.log(res, 10)} y el valor es {res}")

El número de dígitos de la solución media es 18.19932910926361 y el valor es 1.5824467664720458e+18
CPU times: user 5min 41s, sys: 952 ms, total: 5min 42s
Wall time: 5min 45s


El resultado final que hemos obtenido ha sido de **18.2 dígitos** en 5min 40s, es decir, **3.4s** por cada ejecucción.